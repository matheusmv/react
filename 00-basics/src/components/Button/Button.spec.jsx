import React from 'react';

import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Button } from '.';

describe('<Button />', () => {
  it("should render 2 buttons with the text 'back' & 'next'", () => {
    const backFn = jest.fn();
    const nextFn = jest.fn();

    render(<Button onClick={backFn} text="back" />);
    render(<Button onClick={nextFn} text="next" />);

    expect.assertions(7);

    const buttons = screen.getAllByRole('button', { name: /back|next/i });

    expect(buttons.length).toBe(2);

    const backButton = buttons[0];
    const nextButton = buttons[1];

    expect(backButton).toBeInTheDocument();
    expect(backButton).toHaveClass('button');
    expect(backButton).toHaveTextContent('back');

    expect(nextButton).toBeInTheDocument();
    expect(nextButton).toHaveClass('button');
    expect(nextButton).toHaveTextContent('next');
  });

  it('should call function on button click', () => {
    const backFn = jest.fn();
    const nextFn = jest.fn();

    render(<Button text="back" onClick={backFn} />);
    render(<Button text="next" onClick={nextFn} />);

    expect.assertions(3);

    const buttons = screen.getAllByRole('button', { name: /back|next/i });

    expect(buttons.length).toBe(2);

    const backButton = buttons[0];
    const nextButton = buttons[1];

    fireEvent.click(backButton);
    expect(backFn).toHaveBeenCalled();

    userEvent.click(nextButton);
    expect(nextFn).toHaveBeenCalled();
  });

  it("should be disabled when 'disabled' is true", () => {
    const backFn = jest.fn();

    render(<Button onClick={backFn} text="back" disabled />);

    expect.assertions(1);

    const button = screen.getByRole('button', { name: /back/i });

    expect(button).toBeDisabled();
  });

  it("should be enabled when 'disabled' is false", () => {
    const backFn = jest.fn();

    render(<Button onClick={backFn} text="back" disabled={false} />);

    expect.assertions(1);

    const button = screen.getByRole('button', { name: /back/i });

    expect(button).toBeEnabled();
  });

  it('should match snapshot', () => {
    const backFn = jest.fn();

    const { container } = render(<Button text="back" disabled={false} onClick={backFn} />);

    expect(container).toMatchSnapshot();
  });
});
