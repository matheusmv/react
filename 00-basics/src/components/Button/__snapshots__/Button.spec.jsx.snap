// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should match snapshot 1`] = `
<div>
  <button
    class="button"
    type="button"
  >
    back
  </button>
</div>
`;
